Add Impeller migration step for camera plugin authors (#11008)
flutter/website: src/content/release/breaking-changes/android-surface-plugins.md

A full example of using this new API can be found in [PR 6989][] for the
`video_player_android` plugin.

## Note on camera previews

If your plugin implements a camera preview, your migration may also require
fixing the rotation of that preview. This is because `Surface`s produced by the
`SurfaceProducer` may not contain the transformation information that Android
libraries need to correctly rotate the preview automatically.

> In order to correct the rotation, you will need to rotate the preview with

Reviewer: Perhaps we should mention using `RotatedBox` to do the rotation?

respect to the camera sensor orientation and the device orientation according
to the equation:

```plaintext

> rotation = (sensorOrientationDegrees - deviceOrientationDegrees * sign + 360) % 360

Reviewer: Should we provide more context on how to retrieve `sensorOrientationDegrees` and `deviceOrientationDegrees`
directly in this document? (instead of pointing people to a GitHub PR)

We provide code samples for other things in this migration guide (see line 59 in this file), so can we do the same for this addition as well? I don't feel string about this, but maybe not everyone writing plugins / reading this doc is okay with going through a potentially long Github PR to find what they need, to migrate.

I would definitely keep the PR link, though, so people have a completed migration to look at as well.

Author: I added some links for further clarification because I agree that the PR isn't enough. We also had to make some Flutter camera plugin-specific fixes in it, so I don't want the focus to be there.

I think I'll leave it here unless others feel strongly because I think camera plugin authors are familiar with the APIs and could make the migration with this information.

A full example of using this new API can be found in [PR 6989][] for the
`video_player_android` plugin.

## Note on camera previews

If your plugin implements a camera preview, your migration may also require
fixing the rotation of that preview. This is because `Surface`s produced by the

> `SurfaceProducer` may not contain the transformation information that Android

Reviewer 2: ```suggestion
`SurfaceProducer` might not contain the transformation information that Android
```

libraries need to correctly rotate the preview automatically.

> In order to correct the rotation, you will need to rotate the preview with

Reviewer 2: ```suggestion
In order to correct the rotation, you need to rotate the preview with
```

respect to the camera sensor orientation and the device orientation according
to the equation:

```plaintext
rotation = (sensorOrientationDegrees - deviceOrientationDegrees * sign + 360) % 360
```

where `deviceOrientationDegrees` is counterclockwise degrees and `sign` is 1 for
front-facing cameras and -1 for back-facing cameras. 

To calculate this rotation,
- Retrieve the sensor orientation degrees by retrieving the value of
[`CameraCharacteristics.SENSOR_ORIENTATION`][].
- Retrieve the device orientation degrees in one of the ways that the
[Android orientation calculation documentation][] details.

> To apply this rotation, you may use a [`RotatedBox`][] widget.

Reviewer 2: ```suggestion
To apply this rotation, you can use a [`RotatedBox`][] widget.
```

> For more information on this calculation see the

Reviewer 2: ```suggestion
For more information on this calculation, check out the
```

[Android orientation calculation documentation][]. For a full example of making

> this fix, see [the PR we used to fix `camera_android_camerax`][].

Reviewer 2: ```suggestion
this fix, check out [this `camera_android_camerax` PR][].
```

[`createSurfaceTexture`]: {{site.api}}/javadoc/io/flutter/view/TextureRegistry.html#createSurfaceTexture()
[`getSurface()`]: {{site.api}}/javadoc/io/flutter/view/TextureRegistry.SurfaceProducer.html#getSurface()

> [`setCallback`]: https://main-api.flutter.dev/javadoc/io/flutter/view/TextureRegistry.SurfaceProducer.html#setCallback(io.flutter.view.TextureRegistry.SurfaceProducer.Callback)

Reviewer 2: ```suggestion
[`setCallback`]: {{site.api}}/javadoc/io/flutter/view/TextureRegistry.SurfaceProducer.html#setCallback(io.flutter.view.TextureRegistry.SurfaceProducer.Callback)
```

> [`CameraCharacteristics.SENSOR_ORIENTATION`]: https://developer.android.com/reference/android/hardware/camera2/CameraCharacteristics#SENSOR_ORIENTATION

Reviewer 2: ```suggestion
[`CameraCharacteristics.SENSOR_ORIENTATION`]: {{site.android-dev}}/reference/android/hardware/camera2/CameraCharacteristics#SENSOR_ORIENTATION
```

> [this logic in our camera plugin]: https://github.com/flutter/packages/blob/d9a6de802e1fa74b377721929bfa6de5716ce6d9/packages/camera/camera_android_camerax/android/src/main/java/io/flutter/plugins/camerax/DeviceOrientationManager.java#L127

Reviewer 2: @camsim99, "this logic in our camera plugin" text doesn't seem to appear in the PR. At least I couldn't find it. If not, can we delete this definition? Or make this a valid link. thx

Author: Oops! Let me remove this.

> [`RotatedBox`]: https://api.flutter.dev/flutter/widgets/RotatedBox-class.html

Reviewer 2: ```suggestion
[`RotatedBox`]: {{site.api}}/flutter/widgets/RotatedBox-class.html
```

> [Android orientation calculation documentation]: https://developer.android.com/media/camera/camera2/camera-preview#orientation_calculation

Reviewer 2: ```suggestion
[Android orientation calculation documentation]: {{site.android-dev}}/media/camera/camera2/camera-preview#orientation_calculation
```

> [the PR we used to fix `camera_android_camerax`]: https://github.com/flutter/packages/pull/7044

Reviewer 2: ```suggestion
[this`camera_android_camerax` PR]: {{site.repo.flutter}}/packages/pull/7044
```
